### ISSUE TITLE
Allow collecting CPU diagnostics in the Fleet diagnostics action handler

### ISSUE THREAD

**Issue by cmacknz** [elastic/elastic-agent#3491]
https://github.com/elastic/elastic-agent/issues/2140 restored the ability of the `elastic-agent diagnostics` command to collect CPU profiles, which is critically important when diagnosing CPU performance problems.

CPU profile collection is optional and is done when the `--cpu-profiles` option is provided like `elastic-agent diagnostics --cpu-profile`.

However, the Fleet diagnostics action was not modified as part of this change so diagnostics requests from Fleet cannot collect CPU profiles. This is the agent side issue to support the optional diagnostics parameter that will be sent from Fleet as part of https://github.com/elastic/kibana/issues/162849.

Requesting CPU diagnostics is done by supply the `AdditionalDiagnosticRequest_CPU` option in the additional diagnostics request. This logic needs to be implemented in the Fleet diagnostics action handler when the to be determined flag to request CPU diagnostics from Fleet is present.

Here is the agent diagnostics action handler today:

https://github.com/elastic/elastic-agent/blob/123ba9ce80c9865f72fa3659b5cafe9b51954f49/internal/pkg/agent/application/actions/handlers/handler_action_diagnostics.go#L124-L138

Here is how the `elastic-agent diagnostics` handler makes calls to the control protocol server to request CPU profiles:

https://github.com/elastic/elastic-agent/blob/123ba9ce80c9865f72fa3659b5cafe9b51954f49/internal/pkg/agent/cmd/diagnostics.go#L81-L101

Here is the corresponding implementation in the agent control protocol server used by the `elastic-agent diagnostics` command showing how to request CPU profiles.

* For the agent itself: 
https://github.com/elastic/elastic-agent/blob/123ba9ce80c9865f72fa3659b5cafe9b51954f49/pkg/control/v2/server/server.go#L186-L208

* For units:
https://github.com/elastic/elastic-agent/blob/123ba9ce80c9865f72fa3659b5cafe9b51954f49/pkg/control/v2/server/server.go#L268-L282

* For components:
https://github.com/elastic/elastic-agent/blob/123ba9ce80c9865f72fa3659b5cafe9b51954f49/pkg/control/v2/server/server.go#L230-L239

**Comment by michel-laterman** [comment]
@cmacknz 

Wouldn't the optional CPU handler that is registered as part of the default handlers with the client collect this?
https://github.com/elastic/elastic-agent-client/blob/28d7d9587e88470c9a2733358d2dc9ca9d0526a0/pkg/client/client_v2.go#L1056-L1065

**Comment by cmacknz** [comment]
Yes the agent today can optionally collect CPU profiles. It is optional because the default profile is 30s per sub-process, so it takes a long time.

What we need is a way for the diagnostics request from Fleet to expose this option. I suppose we could just start doing it by default for every diagnostics collected from Fleet, but there are instances with many inputs where doing this could take many minutes for each diagnostic bundle.

**Comment by cmacknz** [comment]
@intxgo The CPU profile option here is meant to trigger collection of the Go runtime's pprof CPU profile type, which by default takes 30s to complete, hence why it is optional. https://pkg.go.dev/runtime/pprof. The majority of diagnostics requests do not request CPU profiles, unless they are specifically attempting to debug a CPU utilization problem.

The CPU action type does imply a longer timeout is required, but right now the timeouts are a bit arbitrary and hard coded into the agent. 

https://github.com/elastic/elastic-agent/blob/8ea0d04402aa62cf745d39ff535bc4ce7f642e23/pkg/component/runtime/manager.go#L981-L988

If endpoint needs a longer timeout than these defaults we can probably come up with a nicer way to handle that. Perhaps we can put it in the spec file.

> By the way, will it be acceptable to parametrize the diagnostics context deadline per OS in endpoint-security.spec.yaml? Also, from Endpoint perspective, we would much more prefer the REQUEST_DIAGNOSTICS action repeated for each file until Endpoint explicitly ACKs no more diag files to be made. This way the context deadline could be shorter per file. Endpoint produces a different daignostic artifacts than the rest of the Beats, and we're going to add even more diag files in the near future.

We don't look at the contents of diagnostics and you can structure them however you want, and include as many results as needed. 

The way we use diagnostics has evolved a bit since the protocol was initially written, but there is a diagnostics response that contains an array of diagnostics results. The name implies they have to do with units in the protocol, but there can actually be as many of them as you want included.

https://github.com/elastic/elastic-agent-client/blob/28d7d9587e88470c9a2733358d2dc9ca9d0526a0/elastic-agent-client.proto#L443-L444

Our Go control protocol client includes the concept of a diagnostic hook for attaching custom results. There is no limit to how many there can be, and they just append things to that results array.

Here's an example in Filebeat attaching a snapshot of the Filebeat input metrics as a JSON file for example:

https://github.com/elastic/beats/blob/491049a2074caf7299041fd63cb8eadc29036592/filebeat/beater/filebeat.go#L138-L146

So the summary would be:
1. We can adjust the timeout for endpoint if needed, I wouldn't depend on or expect the CPU profile option to increase the timeout if needed.
2. You can included however many arbitrary `ActionDiagnosticUnitResult` in an `ActionResponse` as you want. We should probably just change the name to `ActionDiagnosticResult` at this point.

**Comment by intxgo** [comment]
> Our Go control protocol client includes the concept of a diagnostic hook for attaching custom results. There is no limit to how many there can be, and they just append things to that results array.

Indeed, but the Endpoint/Beat has to respond only once with the array withing the given deadline context. Only once per each UNIT, so up to 3 times. My thought built on that, just imagine instead of returning an array, the beat/endpoint will return one file assuming agent is going to repeat (with fresh context deadline) the request, and so on until the endpoint returns just empty ack to signal "no more diags for that UNIT/COMPONENT". This sounds more flexible for me.

I see https://github.com/elastic/beats/blob/491049a2074caf7299041fd63cb8eadc29036592/filebeat/beater/filebeat.go#L138-L146 with CPU metrics the context is longer, but it's still absolute, not anything like "default context deadline" + 30s

**Comment by cmacknz** [comment]
> . My thought built on that, just imagine instead of returning an array, the beat/endpoint will return one file assuming agent is going to repeat (with fresh context deadline) the request, and so on until the endpoint returns just empty ack to signal "no more diags for that UNIT/COMPONENT". This sounds more flexible for me.

This would be a specific addition to the protocol for endpoint, since the current mechanism works for all other use cases right now. We could change or add to the protocol, but IMO only if the current mechanism is not capable of solving your problem and that isn't clear to me yet.

Open a separate issue and we can continue the discussion, we are getting off track of the current issue's original purpose.
